File: utils/Operators/UPDATE_THERMODYNAMIC_PROPERTIES.m
function s = UPDATE_THERMODYNAMIC_PROPERTIES(s, chemistry)
% UPDATE_THERMODYNAMIC_PROPERTIES  Compute pressure and temperature from conserved variables.
%
%   s = UPDATE_THERMODYNAMIC_PROPERTIES(s, chemistry)
%
%   Evaluates pressure using p = rho * e * (gamma* - 1) and temperature
%   using T = e * energy_factor / cv_star. 
%
%   Inputs:
%       s  - Solution struct containing:
%                   .var.rho, .var.rho_u, .var.rho_v, .var.rho_E - Conserved variables
%                   .var.gamma_star      - Ratio of specific heats field
%                   .var.cv_star         - Specific heat at constant volume
%       chemistry - Chemistry model struct (reserved for future use)
%
%   Outputs:
%       s  - Updated s struct with:
%                   .var.p - Pressure field (non-dimensional, rho * U^2)
%                   .var.T - Temperature field (non-dimensional, U^2 / cv_infty)
%
%   Notes:
%       - Internal energy: e = (rho_E - 0.5*(rho_u^2 + rho_v^2)/rho) / rho
%
% Part of: Hypersonics Stability MATLAB Solver - Operators Module

    %% Compute internal energy, pressure, and temperature
    rho = s.var.rho;
    e = (s.var.rho_E - (s.var.rho_u.^2 + s.var.rho_v.^2) ./ rho / 2) ./ rho;
    s.var.p = rho .* e .* (s.var.gamma_star - 1);               % Non-dimensional rho * U^2
    s.var.T = e ./ s.var.cv_star;                               % Non-dimensional U^2 / cv_infty

    %% Wall boundary condition at eta0, keep pressure smooth, just update temperature for isothermal walls
    switch s.boundary_conditions.boundary_eta0.name
        case 'no_slip_isothermal'
            s.var.p(:, 1) = 2 * s.var.p(:, 2) - s.var.p(:, 3); % Interpolate pressure
    end
end

File: utils/Operators/UPDATE_THERMODYNAMIC_PROPERTIES.jl
function UPDATE_THERMODYNAMIC_PROPERTIES(s::Dict{String, Any}, chemistry::Dict{String, Any})
# UPDATE_THERMODYNAMIC_PROPERTIES - Compute pressure and temperature from conserved variables.
#
# Evaluates pressure using p = rho * e * (gamma* - 1) and temperature
# using T = e / cv_star.
#
# Part of: Hypersonics Stability Julia Solver - Operators Module

    ## Extract dict refs
    var = s["var"]::Dict{String, Any}
    bc  = s["boundary_conditions"]::Dict{String, Any}

    ## Compute internal energy, pressure, and temperature
    rho         = var["rho"]::Matrix{Float64}
    rho_E       = var["rho_E"]::Matrix{Float64}
    rho_u_arr   = var["rho_u"]::Matrix{Float64}
    rho_v_arr   = var["rho_v"]::Matrix{Float64}
    gamma_star  = var["gamma_star"]::Matrix{Float64}
    cv_star     = var["cv_star"]::Matrix{Float64}

    ## Compute internal energy, pressure, and temperature via fused broadcasting
    e = @. (rho_E - (rho_u_arr^2 + rho_v_arr^2) / rho / 2) / rho
    var["p"] = @. rho * e * (gamma_star - 1)               # Non-dimensional rho * U^2
    var["T"] = @. e / cv_star                                # Non-dimensional U^2 / cv_infty

    ## Wall boundary condition at eta0, keep pressure smooth, just update temperature for isothermal walls
    bc_eta0 = bc["boundary_eta0"]["name"]
    if bc_eta0 == "no_slip_isothermal"
        p = var["p"]::Matrix{Float64}
        @views @. p[:, 1] = 2 * p[:, 2] - p[:, 3]  # Interpolate pressure
    end

    return s
end
